### PR TITLE
[skip ci]Feature/pr gate pipeline

### PR DESCRIPTION
**Pull request by changev**
1. Modify pr_parser to generate manifest file
2. add pr gate manifest template
3. Modify ```class manifest``` to support pr_parser

@panpan0000 @PengTian0

### CODE REVIEW

File: build-release-tools/lib/manifest.py [hunk 1]
```diff
@@ -324,6 +323,97 @@ def check_commit_changed(repo, repo_url, branch, commit):
...
+            # If repo has "branch", compare "commit-id" in repo with the argument commit
+            # only when "branch" is the same with argument branch.
+
+            if 'commit-id' in repo:
+                if repo['commit-id'] != commit:
+                    return False
+            
+            # Exits with 1 if repo doesn't have "branch" and "commit-id"
+            elif 'branch' not in repo:
+                    raise KeyError("Neither commit-id nor branch is set for repository {0}".format(repo['repository']))
```
Comment (anhou) [inline build-release-tools/lib/manifest.py:348]: just confirm if other places are catching this error, from the function description, there's no output for exception

File: build-release-tools/lib/manifest.py [hunk 2]
```diff
@@ -324,6 +323,97 @@ def check_commit_changed(repo, repo_url, branch, commit):
...
+        :param repo: an repository entry of member _repositories or _downstream_jobs
+        :param repo_url: the url of the repository
+        :param branch: the branch of the repository
+        :param commit: the commit id of the repository
+        :return: True when the branch is different with argument branch
+                 and the url and commit is the same with the arguments repo_url, commit;
+                 otherwise, False
+        """
+
+        if (repo['repository'] == repo_url):
```
Comment (anhou) [inline build-release-tools/lib/manifest.py:338]: how about case !=,  by the way, if there are several 'if else' pairs' , guard sentences could be used

File: pr_parser.py [hunk 3]
```diff
@@ -217,100 +158,119 @@ def get_all_related_prs(self, repo, sha, pull_id, actual_commit):
...
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--change-url",
+                        help="Url of the triggered pr",
+                        action="store")
+    parser.add_argument("--target-branch",
+                        help="The target branch of the pr for the named repo",
+                        required=True,
+                        action="store")
+    parser.add_argument("--ghtoken",
```
Comment (anhou) [inline pr_parser.py]: why not named '--github-token' just like other parameters styles

File: pr_parser.py [hunk 4]
```diff
@@ -217,100 +158,119 @@ def get_all_related_prs(self, repo, sha, pull_id, actual_commit):
...
+        According to the self.__pr_connectivity_map and self.__pr_list
+        to find those under_test_prs.
+        under_test_pr: 
+            1. root pr(which trigger this build) is under_test
+            2. pr that diconnected with under_test_pr is under_test_pr
+        """
+        # the final filtered results
+        # under_test_prs to be parsed to find related under_test_prs
+        tmp_under_test_prs = [self.__pr_list[0]]
+        # wich is the under_test_prs after parsing
```
Comment (anhou) [inline pr_parser.py]: typo 'which'

File: pr_parser.py [hunk 5]
```diff
@@ -217,100 +158,119 @@ def get_all_related_prs(self, repo, sha, pull_id, actual_commit):
...
+        under_test_prs = [self.__pr_list[0]]
+        # prs haven been parsed, for preventing infinite loop
+        parsed_prs = []
+        while tmp_under_test_prs:
+            parsed_prs.extend(tmp_under_test_prs)
+            # the next tmp_under_test_prs
+            tmp_tmp_under_test_prs = []
```
Comment (anhou) [inline pr_parser.py]: what these variables 'tmp_tmp_under_test_prs' 'tmp_under_test_prs' 'under_test_prs ''  used for,  could their intention be reflected in their names?

File: pr_parser.py [hunk 6]
```diff
@@ -217,100 +158,119 @@ def get_all_related_prs(self, repo, sha, pull_id, actual_commit):
...
-                for repo_name in ['ON_HTTP', 'ON_TASKGRAPH']:
-                    if properties_dic.has_key(repo_name):
-                        properties_dic[repo_name] = "false"
-            
-            #if on-core exists,all unit-test will be done in on-core sub job so disable them here 
-            if properties_dic.has_key('ON_CORE'):
-                for repo_name in ['ON_HTTP', 'ON_DHCP_PROXY', 'ON_TASKGRAPH', 'ON_TASKS', 'ON_TFTP', 'ON_SYSLOG']:
-                    if properties_dic.has_key(repo_name):
-                        properties_dic[repo_name] = "false"
+        gh = self.__gh
```
Comment (anhou) [inline pr_parser.py]: what gh stands for is a little confused for me.

File: pr_parser.py [hunk 7]
```diff
@@ -217,100 +158,119 @@ def get_all_related_prs(self, repo, sha, pull_id, actual_commit):
...
+
+        # wrap with pr
+        repo_url_list = [repo["repository"] for repo in manifest.repositories]
+        for pr in all_prs:
+            repo, sha1, _ = pr
+            repo_url = "https://github.com/{0}.git".format(repo)
+            # uniform the repo_url case, make sure the url is completely consistent with repo in the manifest
+            repo_url = [url for url in repo_url_list if url.lower()== repo_url][0]
+            if repo in under_test_prs:
+                manifest.update_manifest(repo_url, "", sha1, True)
```
Comment (anhou) [inline pr_parser.py]: could it be  manifest.update_manifest(repo_url, "", sha1,  repo in under_test_prs) ?

File: pr_parser.py (hunk 3 above)
Comment (changev) [reply pr_parser.py]: a conventional rule inherited since onrack ci :)

File: build-release-tools/lib/manifest.py (hunk 1 above)
Comment (changev) [reply build-release-tools/lib/manifest.py:348]: user of this class should handle with the exception.
or just end their running scripts when exception occurs.

File: pr_parser.py (hunk 6 above)
Comment (changev) [reply pr_parser.py]: github

File: build-release-tools/lib/manifest-pr-gate.json [hunk 8]
```diff
@@ -0,0 +1,57 @@
+{
+    "build-name": "rackhd-devel", 
+    "build-requirements": "xenial", 
+    "downstream-jobs": [], 
+    "repositories": [
```
Comment (PengTian0) [inline build-release-tools/lib/manifest-pr-gate.json:5]: Please add the repository: https://github.com/RackHD/ucs-service

**Review by anhou — APPROVED** [review]

**Comment by PengTian0** [comment]
:+1:

**Pull request merged**